### PR TITLE
refactor: deduplicate integer serialization in RollingBloom benchmark

### PR DESCRIPTION
**Pull request by phyBrackets**
refactor: deduplicate integer serialization in RollingBloom benchmark and i think we have to follow don't repeat yourself pattern here! I don't think that this much affect the memory and efficiency because it is still in constant time . But surely it reduces some readability of code which is not the big headache for sure.

### CODE REVIEW

**Review by theStack — COMMENTED** [review]
Seems like you could also simply use the helpers `WriteLE32` and `WriteBE32` from the `<crypto/common.h>` header in this case to avoid manual bit-fiddling:
https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L44-L48

https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L77-L81

As PR and commit subject I'd suggest using something more concrete like "refactor: deduplicate integer serialization in RollingBloom benchmark", rather than a (rhetorical) question.

**Comment by phyBrackets** [comment]
> Seems like you could also simply use the helpers `WriteLE32` and `WriteBE32` from the `<crypto/common.h>` header in this case to avoid manual bit-fiddling:
> 
> https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L44-L48

> https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L77-L81
> 
> As PR and commit subject I'd suggest using something more concrete like "refactor: deduplicate integer serialization in RollingBloom benchmark", rather than a (rhetorical) question.

**Comment by phyBrackets** [comment]
Thanks for the suggestion. I update the following.
@theStack

**Comment by phyBrackets** [comment]
would you tell me @theStack why is there some checks were not successful? and how can i resolve it?

File: src/bench/rollingbloom.cpp [hunk 1]
```diff
@@ -13,16 +13,13 @@ static void RollingBloom(benchmark::Bench& bench)
     uint32_t count = 0;
     bench.run([&] {
         count++;
-        data[0] = count & 0xFF;
-        data[1] = (count >> 8) & 0xFF;
-        data[2] = (count >> 16) & 0xFF;
-        data[3] = (count >> 24) & 0xFF;
+        for(int i = 0; i <= 3; i++){
```
Comment (sipa) [inline src/bench/rollingbloom.cpp:16]: This is just `WriteLE32(data, count);`.

File: src/bench/rollingbloom.cpp [hunk 2]
```diff
@@ -13,16 +13,13 @@ static void RollingBloom(benchmark::Bench& bench)
...
+            data[i] = (count >> 8*i) & 0xFF;
+        }
         filter.insert(data);
-
-        data[0] = (count >> 24) & 0xFF;
-        data[1] = (count >> 16) & 0xFF;
-        data[2] = (count >> 8) & 0xFF;
-        data[3] = count & 0xFF;
+        for(int i = 3; i >= 0; i--){
```
Comment (sipa) [inline src/bench/rollingbloom.cpp:20]: This is just `WriteBE32(data, count);`.

**Comment by fanquake** [comment]
Once you've addressed the suggestions, make sure you squash your changes to a single commit, with a proper commit message.

**Comment by phyBrackets** [comment]
> Once you've addressed the suggestions, make sure you squash your changes to a single commit, with a proper commit message.


@fanquake will take care of that. Thanks.

File: src/bench/rollingbloom.cpp (hunk 2 above)
Comment (phyBrackets) [reply src/bench/rollingbloom.cpp:20]: so, should i change the following to just a function call by adding `common.h` header ?

**Comment by phyBrackets** [comment]
> Seems like you could also simply use the helpers `WriteLE32` and `WriteBE32` from the `<crypto/common.h>` header in this case to avoid manual bit-fiddling:
> 
> https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L44-L48
> 
> https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L77-L81
> 
> As PR and commit subject I'd suggest using something more concrete like "refactor: deduplicate integer serialization in RollingBloom benchmark", rather than a (rhetorical) question.


Isn't this bit inefficient because we are actually dropping the whole content of  `common.h` ?

File: src/bench/rollingbloom.cpp (hunk 2 above)
Comment (sipa) [reply src/bench/rollingbloom.cpp:20]: Yeah, better than duplicating the logic here/

**Comment by phyBrackets** [comment]
> > Seems like you could also simply use the helpers `WriteLE32` and `WriteBE32` from the `<crypto/common.h>` header in this case to avoid manual bit-fiddling:
> > https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L44-L48
> > 
> > https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/crypto/common.h#L77-L81
> > 
> > As PR and commit subject I'd suggest using something more concrete like "refactor: deduplicate integer serialization in RollingBloom benchmark", rather than a (rhetorical) question.
> 
> Isn't this bit inefficient because we are actually dropping the whole content of `common.h` ?


Hey @sipa what's your views on this before making a function call ?

**Comment by sipa** [comment]
I don't understand your question. Using ReadLE32/WriteLE32 is almost certainly more efficient than reimplementing it yourself (they're written in a way that the compiler can easily optimize out). Not that ~nanosecond level performance is relevant here...

**Comment by phyBrackets** [comment]
oh , okay ! Then it's fine I think.

**Comment by phyBrackets** [comment]
Sorry for the direct commit , need to do some more changes.

**Comment by phyBrackets** [comment]
Hey, any update on this?

**Comment by MarcoFalke** [comment]
I haven't looked at the changes, but this can't be merged unless squashed.

Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

**Comment by phyBrackets** [comment]
Cool, will do that soon. Thanks!

**Comment by phyBrackets** [comment]
Hey, I'm having some problem with squashing the commits. Should I make new PR with final changes? Will close this and refer to this PR! If you would allow?

**Comment by fanquake** [comment]
>  Should I make new PR with final changes?

No. Please don't open new PRs for the same changes. You might need to take another look through: https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits/.

**Comment by phyBrackets** [comment]
Hey @fanquake , I tried to rebase/squash , I have some problems with branches I think , there is branch conflict . I really don't want to open a new PR but it seems the only option , by the way if you would allow, I'll open that and set the reference to this PR , and will close this pr ?

**Comment by sipa** [comment]
@phyBrackets It's probably worthwhile learning how to resolve these issues anyway, if you want to continue contributing, whether you do that for this PR or a future one.

File: src/bench/rollingbloom.cpp [hunk 3]
```diff
@@ -5,24 +5,20 @@
 
 #include <bench/bench.h>
 #include <common/bloom.h>
+#include <crypto/common.h>
 
 static void RollingBloom(benchmark::Bench& bench)
 {
     CRollingBloomFilter filter(120000, 0.000001);
     std::vector<unsigned char> data(32);
+    unsigned char* dataPtr = &data[0];
```
Comment (MarcoFalke) [inline src/bench/rollingbloom.cpp:14]:
```suggestion
    unsigned char* dataPtr = data.data();
```

Shouldn't this use the C++11 member function?

Comment (phyBrackets) [reply src/bench/rollingbloom.cpp]: Yeah, resolve this with the new commit.

**Comment by phyBrackets** [comment]
> @phyBrackets It's probably worthwhile learning how to resolve these issues anyway, if you want to continue contributing, whether you do that for this PR or a future one.

Hey @sipa , I'm really trying and searching around how to resolve this issue but I'm not able to , actually when I' trying to run `git rebase -i HEAD~7` , this only shows the first 4 commits.

**Comment by MarcoFalke** [comment]
Your branch contains merge commits, which can't be squashed in a rebase.

Something like this might work (untested!!):

```
git checkout patch-1                                       # Verify this is on commit aa91fd400a6a23feeacdff0d2adacd18dd313e47
git fetch origin 5d254a234d8c1569b0161264cc6d5d8d0ce0d864  # Fetch the latest master
git merge 5d254a234d8c1569b0161264cc6d5d8d0ce0d864         # Merge the latest master
git reset --soft 5d254a234d8c1569b0161264cc6d5d8d0ce0d864  # Discard commit history, but keep all changes staged
git commit --message="bla bla"                             # Create a commit with all staged changes

**Comment by phyBrackets** [comment]
I tried this , it seems more mess tho.

**Comment by MarcoFalke** [comment]
After the steps in https://github.com/bitcoin/bitcoin/pull/24088#issuecomment-1045967132 you need to force push, not merge

**Comment by phyBrackets** [comment]
Thank you sm @MarcoFalke . It seems like it works fine.

**Review by theStack — APPROVED** [review]
Code-review ACK e1fcf8dc148615842341726c6aa35bb00c35f1d0

Nit: I don't think the `dataPtr` pointer is really needed here, you could just call the serialization helper functions with `data.data()` as first parameter each.

**Comment by phyBrackets** [comment]
Cool, but I think it is more readable through this . But yeah if this affects on unnecessary memory use. We can remove that. Let me know your further take on this.

**Comment by laanwj** [comment]
> Nit: I don't think the dataPtr pointer is really needed here, you could just call the serialization helper functions with data.data() as first parameter each.

+1. The assignment is unnecessary. I doubt it affects memory use in any significant way but I see no reason not to write it shorter if you're touching this code anyway.

**Comment by fanquake** [comment]
I've just gone ahead and fixed this up in #24784, given this is straightforward, and likely been open long enough.

**Pull request closed without merging**